File: server/src/services/taskRouter.js
// Cheap regex-based router, no LLM call — decides BEFORE any retrieval/web
// search which agent handles this turn. Priority order enforced here:
// Uploaded Image > Uploaded Document > Coding > Memory/RAG > Web Search. If
// an image or document is attached, web-search-enabled general path must
// never run.
const RESUME_RE = /\bresume\b|\bcv\b/i;
const ATS_RE = /\bats\b.*(score|calculate|rate)|calculate.*\bats\b/i;
const ANALYZE_RE = /\b(review|analyze|analyse|summarize|summarise|check|evaluate|critique)\b/i;
const COMPARE_RE = /\b(compare|comparison|diff|difference|differences|contrast)\b.*\b(document|doc|file|pdf|these|them|both)\b|\b(document|doc|file)s?\b.*\b(compare|differ|contrast)\b/i;

// Coding questions skip the whole research->plan->review pipeline (RAG,
// web search, multi-source evidence) — none of that helps "write a
// mergesort in Python", it only adds 1-3min of latency for zero benefit.
// Matches explicit language/task asks ("write java code", "python
// function", "leetcode", "sql query", "react component") rather than every
// mention of a language name, to avoid false-positiving on "what is Python
// used for" (a factual question, not a code-gen request).
const CODE_LANGS = /(java|python|javascript|typescript|c\+\+|c#|go|golang|rust|ruby|php|sql|html|css|kotlin|swift|react|node\.?js|express)/i;
const CODE_VERBS = /\b(write|give|generate|create|implement|build|code|fix|debug|refactor|optimi[sz]e)\b/i;
const CODE_NOUNS = /\b(code|function|algorithm|script|program|snippet|class|component|endpoint|api|query|regex)\b/i;
const CODE_DOMAIN_RE = /\b(leetcode|dsa|data structure|hackerrank|codeforces|merge sort|quick sort|binary search|two sum|fibonacci)\b/i;

export function isCodingQuestion(query) {
  if (CODE_DOMAIN_RE.test(query)) return true;
  return CODE_VERBS.test(query) && (CODE_NOUNS.test(query) || CODE_LANGS.test(query));
}

export function classifyTask(query, { hasFiles, hasImages, fileCount = 0 } = {}) {
  if (hasImages) return { type: 'vision' };
  if (!hasFiles && isCodingQuestion(query)) return { type: 'coding' };
  if (!hasFiles) return { type: 'general' };

  if (fileCount >= 2 && COMPARE_RE.test(query)) return { type: 'document_comparison' };
  if (ATS_RE.test(query)) return { type: 'ats' };
  if (RESUME_RE.test(query) && ANALYZE_RE.test(query)) return { type: 'resume_analysis' };
  if (ANALYZE_RE.test(query)) return { type: 'document_analysis' };
  if (isCodingQuestion(query)) return { type: 'coding' };

  return { type: 'general' };
}

File: server/app/services/task_router.py
import re

CODE_LANGS = re.compile(
    r"(java|python|javascript|typescript|c\+\+|c#|go|golang|rust|ruby|php|sql|html|css|kotlin|swift|react|node\.?js|express)",
    re.I,
)
CODE_VERBS = re.compile(r"\b(write|give|generate|create|implement|build|code|fix|debug|refactor|optimi[sz]e)\b", re.I)
CODE_NOUNS = re.compile(r"\b(code|function|algorithm|script|program|snippet|class|component|endpoint|api|query|regex)\b", re.I)
CODE_DOMAIN_RE = re.compile(
    r"\b(leetcode|dsa|data structure|hackerrank|codeforces|merge sort|quick sort|binary search|two sum|fibonacci)\b",
    re.I,
)


def is_coding_question(query: str) -> bool:
    if CODE_DOMAIN_RE.search(query):
        return True
    if CODE_LANGS.search(query) and CODE_NOUNS.search(query):
        return True
    return bool(CODE_VERBS.search(query) and (CODE_NOUNS.search(query) or CODE_LANGS.search(query)))


RESEARCH_RE = re.compile(r"\b(research|in depth|comprehensive|thorough|literature|survey of|state of the art)\b", re.I)
BIOGRAPHY_RE = re.compile(r"\bwho (is|was)\b|\bbiography\b|\bborn (in|on)\b|\blife of\b", re.I)
NEWS_RE = re.compile(r"\b(latest|breaking|today'?s|this week'?s|recent)\b.*\b(news|update|event)s?\b|\bwhat'?s happening\b", re.I)
MATH_RE = re.compile(r"\b(solve|calculate|compute)\b.*\b(equation|integral|derivative|matrix|sum|expression)\b|[0-9]+\s*[-+*/^]\s*[0-9]+", re.I)
MEDICAL_RE = re.compile(r"\b(symptom|diagnos|disease|medication|dosage|treatment|patient|clinical)\w*\b", re.I)
LEGAL_RE = re.compile(r"\b(contract|statute|lawsuit|liability|plaintiff|defendant|clause|jurisdiction|legal(ly)?)\b", re.I)


def classify_topic(query: str) -> str:
    """Lightweight heuristic topic-type classifier (keyword/regex based).

    Distinct from `classify_task`, which decides *which agent handles this
    turn* (coding vs. general). This decides the *subject-matter
    category* of a query — used for routing to research/biography/news/
    math/medical/legal specialists and for confidence scoring (contested
    categories in confidence_engine.py)."""
    q = query or ""
    if is_coding_question(q):
        return "coding"
    if MATH_RE.search(q):
        return "math"
    if MEDICAL_RE.search(q):
        return "medical"
    if LEGAL_RE.search(q):
        return "legal"
    if BIOGRAPHY_RE.search(q):
        return "biography"
    if NEWS_RE.search(q):
        return "news"
    if RESEARCH_RE.search(q):
        return "research"
    return "chat"


def classify_task(query: str, has_files: bool = False, has_images: bool = False, file_count: int = 0) -> dict:
    """Priority order: Coding > Memory/RAG > Web Search.

    Uploaded documents/images are handled by the general RAG retrieval path
    (via the Research agent), not a dedicated per-file-type agent."""
    if is_coding_question(query):
        return {"type": "coding"}

    return {"type": "general"}
